Add SOCKS5 proxy support for kubectl exec

xens opened the pull request:
:warning: This PR if the continuation of the work initially started here https://github.com/kubernetes/kubernetes/pull/84205 that seems to be abandoned by its author.

I still need to rework the tests as mentioned by @liggitt here https://github.com/kubernetes/kubernetes/pull/84205#discussion_r509326938, when it'll be done I'll remove the draft status. 

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it

The PR permits using kubectl commands like `kubectl exec` or `kubectl port-forward` via a SOCKS5 proxy. The PR adds support for SOCKS5 proxies, but does not change HTTP or HTTPS behavior.

Test coverage provided for socks5 with armon/go-socks5 server.

#### Does this PR introduce a user-facing change ?

```release-note
Adds support for kubectl commands (`kubectl exec` and `kubectl port-forward`) via a SOCKS5 proxy. 
```

caesarxuchao reviewed (commented):
Made an initial pass.

cc @liggitt  @TGPSKI @Patrick0308

caesarxuchao reviewed (commented):
Added a few new comments. Thanks for the patience!

xens commented:
@caesarxuchao @Patrick0308 all the comments should be resolved, ready for a new review :)

caesarxuchao commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -384,6 +403,258 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+			for _, k := range sets.StringKeySet(testCases).List() {
+				testCase := testCases[k]
+				t.Run(k, func(t *testing.T) {
+					server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
```
nit: can we extract this SPDY server construction into a standalone function to reduce code duplicate? Both SOCK5 tests and the existing tests can call this server construction function.

xens replied:
Done!

caesarxuchao commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -384,6 +403,258 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+						default:
+							return false
+						}
+					}
```
nit: will the following work? It's a little simpler.

```golang
closed := make(chan struct{})
isClosed := func() bool {
	select {
		case v := <-closed:
			return true
		default:
			return false
	}
}
```

xens replied:
done

caesarxuchao commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -384,6 +403,258 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+						}
+
+						if err := proxyHandler.ServeConn(conn); err != nil && !testCase.shouldError {
+							if isClosed() {
```
nit: can you double check if we need to check `isClosed()` here? I think checking the l.Accept() error case is enough.

xens replied:
Doesn't work without it, not sure why, leaving it for now.

caesarxuchao commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -384,6 +403,258 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+						if expectedSocks5AuthUser != authUser["Username"] {
+							t.Fatalf("%s: Socks5 Proxy authorization user unexpected, got %q, expected %q", k, authUser["Username"], expectedSocks5AuthUser)
+						}
+
```
nit: remove the new line.

xens replied:
done

caesarxuchao commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -384,6 +403,258 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+					var proxierCalled bool
+					var proxyCalledWithHost string
+					var proxyCalledWithAuth bool
+					var proxyCalledWithAuthHeader string
```
proxyCalledWithAuth and proxyCalledWithAuthHeader are not really used in the SOCKS5 tests. Can you double check if I'm right and remove the relevant code?

xens replied:
you're right, adapted the test

caesarxuchao commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -384,6 +403,258 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+							t.Fatalf("%s: Socks5 Proxy authorization user unexpected, got %q, expected %q", k, authUser["Username"], expectedSocks5AuthUser)
+						}
+
+					} else {
```
In the `else` block, we should verify with `interceptor.GetAuthContext()`, rather than the `proxyCalledWithAuth` or  `proxyCalledWithAuthHeader`, which are not populated in the SOCK5 tests.

xens replied:
Done

caesarxuchao reviewed (commented):
A few more nits.

Let's let the sig-auth reviewers take a look as well, since they were involved in the previous attempts of adding SOCK5 supports.

/assign @liggitt @mikedanese

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -163,6 +164,18 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
+		return s.dialWithHttpProxy(req, proxyURL)
+	}
+
+	return nil, fmt.Errorf("proxy URL scheme not supported: %s", proxyURL.Scheme)
```
IIRC before it never failed before, can this break something?

xens replied:
If I understand correctly your question no it shouldn't it has been tested with both scheme and it's working as it should. Maybe the maintainer can confirm ?

xens commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -384,6 +403,258 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+			testCases := map[string]struct {
+				proxyAuth              *url.Userinfo
+				serverConnectionHeader string
+				serverUpgradeHeader    string
```
fixed!

xens commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -384,6 +403,258 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+func TestRoundTripSocks5AndNewConnection(t *testing.T) {
+	for _, redirect := range []bool{false, true} {
+		t.Run(fmt.Sprintf("redirect = %t", redirect), func(t *testing.T) {
+			localhostPool := x509.NewCertPool()
```
We've removed it.

xens commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -17,24 +17,43 @@ limitations under the License.
+	"k8s.io/apimachinery/pkg/util/sets"
 )
 
-// be sure to unset environment variable https_proxy (if exported) before testing, otherwise the testing will fail unexpectedly.
```
We've took back the tests from master and created a test dedicated to the socks5 feature as you suggested and the line was removed accidentally, however we've added it back above the socks5 test.

xens commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -194,9 +209,57 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
+		return nil, err
+	}
+
+	// From documentation https://pkg.go.dev/golang.org/x/net/proxy#Dialer "Custom dialers should also implement ContextDialer."
```
ok done.

liggitt commented on build/tools.go:
```diff
@@ -17,10 +17,12 @@ See the License for the specific language governing permissions and
 
 import (
+	// build script dependencies
+	_ "github.com/go-bindata/go-bindata/go-bindata"
```
this shouldn't be needed any more

xens replied:
It was a leftover from the previous PR, I removed the vendor folder and all the references inside build/tools + go.mod and go.sum.

liggitt commented on go.mod:
```diff
@@ -170,8 +172,10 @@ replace (
 	github.com/ajstarks/svgo => github.com/ajstarks/svgo v0.0.0-20180226025133-644b8db467af
 	github.com/antihax/optional => github.com/antihax/optional v1.0.0
 	github.com/armon/circbuf => github.com/armon/circbuf v0.0.0-20150827004946-bbbad097214e
+	github.com/armon/consul-api => github.com/armon/consul-api v0.0.0-20180202201655-eb2c6b5be1b6
```
where did this dep come from?

xens replied:
left-overs from the old PR, fixed thanks !

xens commented:
@caesarxuchao ready for a next round of review :)

xens commented:
Rebased, ready for reviews :smiley:

caesarxuchao commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -243,142 +300,337 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+						case err != nil && testCase.shouldError:
+							return true
+						case err == nil && testCase.shouldError:
+							t.Fatalf("%s: test should have raised an error but didn't", k)
```
nit: no need to print "k" in the error message anymore.

There are a few references to `k` in this test. Can you remove all of them?

xens replied:
done

caesarxuchao reviewed (commented):
This mostly lgtm now with a few minor comments.

Thanks for the patience!

xens commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -243,142 +300,337 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+						}, interceptor)
+					}
+
+					closed := make(chan struct{}, 1)
```
ok followed your advice !

xens commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -243,142 +300,337 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+			}
+
+			for _, k := range sets.StringKeySet(testCases).List() {
+				testCase := testCases[k]
```
done

xens commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -17,24 +17,81 @@ limitations under the License.
+	return int(i.Authorization.Method), i.Authorization.Payload
+}
+
+type ServerHandlerConfig struct {
```
done

xens commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -17,24 +17,81 @@ limitations under the License.
+
+func (i *Interceptor) GetAuthContext() (int, map[string]string) {
+	return int(i.Authorization.Method), i.Authorization.Payload
+}
```
done

xens commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -17,24 +17,81 @@ limitations under the License.
+	}
+}
+
 func TestRoundTripAndNewConnection(t *testing.T) {
```
done

xens commented:
Ready for a next round of reviews :)

xens commented:
@caesarxuchao sorry to chase you but I'd love to move forward with this PR, let me know if there's anything to do left.

xens commented:
:pray: :)

Patrick0308 commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -194,9 +210,62 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
+		return nil, errors.New("SOCKS5 Dialer must implement ContextDialer")
+	}
+
+	proxyDialConn, err := contextDialer.DialContext(req.Context(), "tcp", targetHost)
+	if err != nil {
+		return nil, err
+	}
+
+	proxyClientConn := httputil.NewProxyClientConn(proxyDialConn, nil)
+
+	rwc, _ := proxyClientConn.Hijack()
+
+	if req.URL.Scheme == "https" {
+		return s.tlsConn(rwc, targetHost)
 	}
```
It's same with the end of dialWithHTTPProxy.

xens replied:
It's true it could be DRYer but it'd add unnecessary complexity to the code so I'm not sure that it's worth it ? Or did you wanted to highlight something else with your comment ?

Patrick0308 replied:
Sorry, forget the my last incorrect comment. There are an another problem. 
```
	proxyClientConn := httputil.NewProxyClientConn(proxyDialConn, nil)
	rwc, _ := proxyClientConn.Hijack()
```
 It don't need to new a http proxy client connection then hijack it. It's redundant.

xens replied:
You're right it's not needed here, corrected !

Patrick0308 reviewed (approved):
LGTM

Patrick0308 commented:
@xens CI failed. Pls fix it.

xens commented:
@Patrick0308 I'll check why the `kubernetes-unit` job is failing, tests are ok on my end I'll come back tomorrow on that one.

For the others `kubernetes-dependencies` `kubernetes-verify` looks like I'm supposed to run some script locally and commit the result  that's correct ? For the `EasyCLA` job I guess that it's due to the initial owner of the PR hasn't signed the CLA ?

xens commented:
@Patrick0308  race in the unit-tests has been fixed as well as `kubernetes-dependencies` `kubernetes-verify` jobs. @liggitt possible to do a review on your end ?

enj commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/BUILD:
```diff
@@ -0,0 +1,60 @@
+package(default_visibility = ["//visibility:public"])
```
We do not use bazel anymore.

xens replied:
Thanks, it was some leftovers from the old PR, I removed the file.

xens commented:
/remove-sig auth

patricknelson commented:
So close! Thanks for this work.  I'm really looking forward to having full `kubectl` functionality over SOCKS proxies. This still appears to be blocked by "EasyCLA — Missing CLA Authorization", see: https://github.com/kubernetes/kubernetes/pull/105632#issuecomment-970244134

> For the others kubernetes-dependencies kubernetes-verify looks like I'm supposed to run some script locally and commit the result that's correct ? For the EasyCLA job I guess that it's due to the initial owner of the PR hasn't signed the CLA ?

Yeah that's weird @xens, it looks like @TGPSKI already signed off on the CLA, per https://github.com/kubernetes/kubernetes/pull/84205#issuecomment-545154335. 

@caesarxuchao @Patrick0308 @liggitt: Any clue why this is blocked by EasyCLA? 😕

xens commented:
EasyCLA is now fixed, we're now missing some `lgtm` label. @liggitt @Patrick0308 @caesarxuchao possible to have your approval ?

xens commented:
@liggitt according to the @k8s-ci-robot we'll need your approval on that one.

xens commented:
:pray: :christmas_tree: :gift: :santa: :snowflake:

caesarxuchao commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -243,142 +296,340 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+
+					client := &http.Client{Transport: spdyTransport}
+
+					checkShouldError := func(err error) bool {
```
nit: For line 555-576, can we use the same code as line 353-365, which doesn't define a standalone `checkShouldError` function.

xens replied:
Agree for both, it should be OK now.

caesarxuchao reviewed (commented):
Two more nits, otherwise LGTM.

xens commented:
/remove-sig auth

xens commented:
Mhhh all the tests are now failing with `cannot find package "golang.org/x/net/proxy"` since my last commit, I'm wondering if something changed in the CI since my last push because I didn't touch that part :thinking:. Help welcome :)

xens commented:
/retest

aojea commented:
> Mhhh all the tests are now failing with `cannot find package "golang.org/x/net/proxy"` since my last commit, I'm wondering if something changed in the CI since my last push because I didn't touch that part thinking. Help welcome :) (@RomainJufer).

that dependency was removed in https://github.com/kubernetes/kubernetes/pull/97252

you have to rebase and update the dependencies with the script `hack/update-vendor.sh`

caesarxuchao commented:
LGTM. This will require a top approver to approve. Let's ask Jordan after the tests are fixed.

xens commented:
/remove-sig auth

xens commented:
rebased on master + ran `hack/update-vendor.sh`, looks like I'm now blocked by this flaky test issue https://github.com/kubernetes/kubernetes/issues/107198

RyanAoh commented:
/retest

xens commented:
Ok looks like we're back on track @caesarxuchao, ready for the final final push! :santa: :gift:

Patrick0308 commented:
Looking forward to using this feature so much

xens commented:
@caesarxuchao possible to involve Jordan to move forward with this PR ?

liggitt commented:
yeah, I'm back after the holiday and this is on my short list for review

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -173,12 +187,14 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
 	}
 
+	proxyReq = *proxyReq.WithContext(req.Context())
+
 	if pa := s.proxyAuth(proxyURL); pa != "" {
 		proxyReq.Header = http.Header{}
 		proxyReq.Header.Set("Proxy-Authorization", pa)
 	}
 
-	proxyDialConn, err := s.dialWithoutProxy(req.Context(), proxyURL)
+	proxyDialConn, err := s.dialWithoutProxy(proxyReq.Context(), proxyURL)
```
Propagating the req context into proxyReq (so it propagates to `proxyClientConn.Do`) looks correct, but it also looks unrelated to adding socks support.

Split this bit into a separate PR we can merge first. That keeps the changes independent, which is helpful for any PR-level operations we need to do like bisecting / backporting / reverting

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -194,10 +210,59 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
-	if req.URL.Scheme != "https" {
-		return rwc, nil
+	if req.URL.Scheme == "https" {
+		return s.tlsConn(rwc, targetHost)
```
plumb `proxyReq.Context()` to `tlsConn` so we can pass it to `tlsConn.HandshakeContext()`

xens replied:
done.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -194,10 +210,59 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
 	}
 
+	if req.URL.Scheme == "https" {
+		return s.tlsConn(proxyDialConn, targetHost)
```
pass `req.Context()` to `tlsConn`

xens replied:
done

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -243,142 +285,348 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+				serverConnectionHeader string
+				serverStatusCode       int
+				shouldError            bool
+			}{
```
I expected to see similar coverage to TestRoundTripAndNewConnection of an https backend with valid/invalid certificates, via the socks proxy ... from what I can see, tls backends via the socks proxy aren't being tested currently

xens replied:
That's a fair point! I've cherry-picked most of the `TestRoundTripAndNewConnection` except that now they are going through the SOCKS5 proxy.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -243,142 +285,348 @@ func TestRoundTripAndNewConnection(t *testing.T) {
-						proxyURL, err := url.Parse(proxy.URL)
+					if testCase.proxyAuth != nil {
+						proxyHandler = socks5Server(&socks5.StaticCredentials{
+							"proxyuser": "proxypasswd",
```
is there a reason these are hardcoded instead of being pulled from testCase.proxyAuth?

xens replied:
This is the server credentials when we instantiate it, that's why they are hardcoded here, the tests then have their own credentials based on the test use-case (if we want to provide valid or invalid credentials) so I think this is correct.

liggitt commented:
Thanks for picking this up. Looks really close, just a few comments.

Go ahead and squash down the commits as well.

Patrick0308 commented:
@xens Do you have time to finish this pr? If not, I can help you.

xens committed:
feat: add missing SOCKS5 features

Goal of this commit is to add some missing features when the
Kubernetes API is accessed through a SOCKS5 proxy. That's for
example the case when port-forwarding is used (`kubectl port-forward`)
or when exec'ing inside a container (`kubectl exec`), with this
commit it'll now be possible to use both.

Signed-off-by: Romain Aviolat <romain.aviolat@kudelskisecurity.com>
Signed-off-by: Romain Jufer <romain.jufer@kudelskisecurity.com>

aojea commented on go.mod:
```diff
@@ -36,7 +37,11 @@ require (
 	github.com/emicklei/go-restful v2.9.5+incompatible
 	github.com/evanphx/json-patch v4.12.0+incompatible
 	github.com/fsnotify/fsnotify v1.4.9
+<<<<<<< HEAD
 	github.com/go-logr/logr v1.2.0
+	github.com/go-bindata/go-bindata v3.1.2+incompatible
+=======
+>>>>>>> fix: remove go-bindata dependency
```
rebase conflict :smile:

xens replied:
omg thanks @aojea nice catch :D

xens commented:
@Patrick0308 I gave this PR some love today and I think we're back on the right track, thanks for your proposal, if you could help my identify why some tests are failing that'd be nice.

@liggitt your latest comments should be resolved, I've rebased on `origin/master` + squashed commits let me know if there's questions.

xens commented:
/remove-sig auth

xens commented:
/retest-required

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -243,142 +297,473 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+					if !proxierCalled {
+						t.Fatal("xpected to use a proxy but proxier in SpdyRoundTripper wasn't called")
+					}
+
```
add a check that the proxy was actually called with the expected host:

```diff
+                                       serverURL, err := url.Parse(server.URL)
+                                       if err != nil {
+                                               t.Fatalf("error creating request: %s", err)
+                                       }
+                                       if proxyCalledWithHost != serverURL.Host {
+                                               t.Fatalf("expected to see a call to the proxy for backend %q, got %q", serverURL.Host, proxyCalledWithHost)
+                                       }
```

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -243,142 +297,473 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+					serverStatusCode:       http.StatusSwitchingProtocols,
+					shouldError:            false,
+				},
+				"proxied https (invalid hostname + InsecureSkipVerify) -> http": {
```
because the socks proxy doesn't have a https variant, we don't need the `"proxied https ... -> ...` tests... there's no difference in the proxy we need to exercise

keep the "with auth" testcases, but rename the tests to clarify they are not testing https proxy attributes:

```
proxied with auth -> http
proxied with auth -> https (invalid hostname)
proxied with auth -> https (valid hostname + RootCAs)
```

xens replied:
Makes sense, I've removed the https tests and rename all the others, it should be more meaningful like this!

liggitt reviewed (commented):
updates look good. two remaining comments on the tests, then lgtm

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -178,7 +192,7 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
 		proxyReq.Header.Set("Proxy-Authorization", pa)
 	}
 
-	proxyDialConn, err := s.dialWithoutProxy(req.Context(), proxyURL)
+	proxyDialConn, err := s.dialWithoutProxy(proxyReq.Context(), proxyURL)
```
or just do 
```
proxyReq = proxyReq.WithContext(req.Context())
```

liggitt replied:
on master, the req.Context() is propagated to proxyReq. See https://github.com/kubernetes/kubernetes/pull/107606

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -194,10 +208,59 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
-	if req.URL.Scheme != "https" {
-		return rwc, nil
+	if req.URL.Scheme == "https" {
+		return s.tlsConn(proxyReq.Context(), rwc, targetHost)
```
same question about context https://github.com/kubernetes/kubernetes/pull/105632/files#r787913440

aojea replied:
solved https://github.com/kubernetes/kubernetes/pull/105632#discussion_r787927859

xens commented:
/remove-sig auth

xens commented:
/retest-required

liggitt commented:
weird... the integration test job seems to be hanging just on this PR... they're still passing on other PRs

liggitt commented:
/retest

liggitt commented:
if this fails again, I think specific investigation is needed why this is causing issues

liggitt commented:
/lgtm
/approve

aojea commented:
:clap:

kfox1111 commented:
Thanks to everyone for helping get this over the finish line! :)

liggitt commented:
looks like the added unit test is flaking - xref https://github.com/kubernetes/kubernetes/issues/107708
